### PR TITLE
Add `auto` option for `config.use_ansi_coloring`

### LINKED ISSUE #11464: Disable term colors when it's not supported
### Related problem

I'm not sure whether it's a bug or a feature request. In some environments where term-colors are not supported, nushell should turn off term-colors.

Below are screenshots of [neovide](https://github.com/neovide/neovide) running cargo and nu. My neovide settings do not support term-coloring. Cargo and Nu both generate colored-outputs on other platforms like alacritty or gnome shell. In neovide, Cargo knows that it cannot use colors, so it outputs plain texts. But Nu still tries to color the outputs, which results hard-to-read texts.

### Describe the solution you'd like

Check whether the current environment supports term-coloring before coloring the output.

### Describe alternatives you've considered

_No response_

### Additional context and details

_No response_

### PR DESCRIPTION
**Pull request by cptpiepmatz**
# Description

In this PR I continued the idea of #11494, it added an `auto` option to the ansi coloring config option, I did this too but in a more simple approach.

So I added a new enum `UseAnsiColoring` with the three values `True`, `False` and `Auto`. When that value is set to `auto`, the default value, it will use `std::io::stdout().is_terminal()` to decided whether to use ansi coloring. This allows to dynamically decide whether to print ansi color codes or not, [cargo does it the same way](https://github.com/rust-lang/cargo/blob/652623b779c88fe44afede28bf7f1c9c07812511/src/bin/cargo/main.rs#L72). `True` and `False` act as overrides to the `is_terminal` check. So with that PR it is possible to force ansi colors on the `table` command or automatically remove them from the miette errors if no terminal is used.

# User-Facing Changes

Terminal users shouldn't be affected by this change as the default value was `true` and `is_terminal` returns for terminals `true` (duh). Non-terminal users, that use `nu` in some embedded way or the engine implemented in some other way (like my jupyter kernel) will now have by default no ansi coloring and need to enable it manually if their environment allows it.

# Tests + Formatting

The test for fancy errors expected ansi codes, since tests aren't run "in terminal", the ansi codes got stripped away.
I added a line that forced ansi colors above it. I'm not sure if that should be the case or if we should test against no ansi colors.

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting

This should resolve #11464 and partially #11847. This also closes #11494.

### CODE REVIEW

**Comment by fdncred** [comment]
Seems pretty straight forward. Thanks!

**Comment by fdncred** [comment]
please fix the conflicts and we can land this

**Comment by cptpiepmatz** [comment]
@fdncred should be fixed

**Comment by fdncred** [comment]
@cptpiepmatz Does this support [NO_COLOR](https://no-color.org/). If it does, I missed it in the code. If not, it would be good to add it.

I'm wondering if we should support `FORCE_COLOR` env var as well? I think this crate supports all this. https://docs.rs/supports-color/latest/supports_color/ I'm not saying we should use it, but it may provide inspiration to make this PR more thorough.

**Comment by cptpiepmatz** [comment]
@fdncred  I made several changes to support environment variables here:

1. The `get` method now requires an `EngineState` to evaluate environment variables. This ensures that we process these variables for every command, meaning each call to `get` will dynamically evaluate them.

2. With the new config behavior, most users will likely have the value set to `auto` by default. When `auto` is used, it respects the `NO_COLOR`, `FORCE_COLOR`, and `CLICOLOR` environment variables. However, users can still explicitly set the value to `true` or `false` in their config to override these variables. Our internal code does this too, for example, to enforce disabling ANSI coloring.

3. Since `get` became a bit more complex, I reorganized some logic in `table` to reduce unnecessary work.

With this new behavior, we should maybe update the docs and add something in the changelog.

File: crates/nu-protocol/src/config/output.rs [hunk 1]
```diff
@@ -24,3 +26,116 @@ impl UpdateFromValue for ErrorStyle {
...
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, IntoValue, Serialize, Deserialize)]
+pub enum UseAnsiColoring {
+    #[default]
+    Auto,
+    True,
+    False,
+}
+
+impl UseAnsiColoring {
+    /// Determines whether ANSI colors should be used.
```
Comment (fdncred) [inline crates/nu-protocol/src/config/output.rs:39]: Great documentation here. Thanks for taking the time to spell out exactly what it does.

File: crates/nu-protocol/src/value/mod.rs [hunk 2]
```diff
@@ -655,6 +655,35 @@ impl Value {
...
+    ///
+    /// For all other, more complex variants of [`Value`], the function cannot determine a
+    /// boolean representation and returns `None`.
+    pub fn as_env_bool(&self) -> Option<bool> {
+        match self {
+            Value::Bool { val: false, .. }
+            | Value::Int { val: 0, .. }
+            | Value::Float { val: 0.0, .. }
+            | Value::Nothing { .. } => Some(false),
+            Value::String { val, .. } if val.as_str() == "" => Some(false),
```
Comment (fdncred) [inline crates/nu-protocol/src/value/mod.rs:679]: I'm wondering if we should support a string 0/1 like `"0"` or `"1"`? Or maybe that's covered somewhere/somehow else?

**Review by fdncred — CHANGES_REQUESTED** [review]
This is exactly what I was asking for. I have a few more comments. Just let me know what you think.

Also, it would be wonderful to have a couple tests for these env var settings, if you think that's possible? I can't remember if our test harness supports such a thing. I also want to test this myself, which I haven't done yet. 

Thanks so much for working on this. I think it'll be a great addition to nushell.

File: crates/nu-protocol/src/value/mod.rs (hunk 2 above)
Comment (cptpiepmatz) [reply crates/nu-protocol/src/value/mod.rs:679]: String "0" should definitely also represent a `false` I just missed that one here. Good point.

**Comment by cptpiepmatz** [comment]
Since `UseAnsiColoring` needed more lines of code than I planned, I moved it into a nother module. I also added tests for `Value::as_env_bool` and `UseAnsiColoring::get`. I don't think I need to test consumers of the `get` method as it simply returns a boolean.

**Comment by cptpiepmatz** [comment]
@fdncred it's green 🎉 

Also imo we should add a label to get this in to the changelog.

**Review by fdncred — COMMENTED** [review]
Sorry to keep hitting you with these "one more things" but I've noticed a couple things.

1. I think we need some text in the [doc_config.nu]( https://github.com/nushell/nushell/blob/38694a9850a08f82a9a887c64df55c9f561e4c61/crates/nu-utils/src/default_files/doc_config.nu#L247-L252)
2. In my testing, I've noticed that it mostly works but there are some edge cases. NO_COLOR=1 but still has some coloring in the banner and the prompt.
image

The banner should be able to be handled in the stdlib perhaps. The prompt is another question. I wonder if the prompts should be ran through ansi-strip or if should be left to the prompt authors? My tendency is to say nushell should remove all coloring. What do you think?

**Comment by cptpiepmatz** [comment]
> Sorry to keep hitting you with these "one more things" but I've noticed a couple things.

I have no problem when the iteration cycle is still fast 😄 

> 1. I think we need some text in the [doc_config.nu](https://github.com/nushell/nushell/blob/38694a9850a08f82a9a887c64df55c9f561e4c61/crates/nu-utils/src/default_files/doc_config.nu#L247-L252)

Definitely, will add some

> 2. In my testing, I've noticed that it mostly works but there are some edge cases. NO_COLOR=1 but still has some coloring in the banner and the prompt.

I'm not sure if other commands properly respect `config.use_ansi_coloring`, I made sure the `table` command does, so we should focus testing with that.

> The banner should be able to be handled in the stdlib perhaps. The prompt is another question. I wonder if the prompts should be ran through ansi-strip or if should be left to the prompt authors? My tendency is to say nushell should remove all coloring. What do you think?

I've mentioned this PR in #14194 and I think we should go in the direction to remove ansi colors only from our own output if not explicitely called with `ansi strip`. So if the banner and prompt use some `ansi` commands, they should respect `config.use_ansi_coloring`.

**Comment by cptpiepmatz** [comment]
@fdncred I updated `doc_config.nu` to accomodate the new changes.

About the prompt and banner, I think this is out of scope of this PR if they respect the config value or not. The `table` command now definitely respects that value and worked in my testing.

**Comment by fdncred** [comment]
I'm fine with landing this as is but I'm not sure we should force 3rd party authors to understand NO_COLOR, FORCE_COLOR, CLI_COLOR. I could see them respecting $env.config.use_ansi_coloring though. But maybe your code is enough with the auto setting. 🤔

**Comment by fdncred** [comment]
Let's give it a whirl. Thanks!

**Comment by cptpiepmatz** [comment]
> I'm fine with landing this as is but I'm not sure we should force 3rd party authors to understand NO_COLOR, FORCE_COLOR, CLI_COLOR. I could see them respecting $env.config.use_ansi_coloring though. But maybe your code is enough with the auto setting. 🤔

Oh, in the sense to check the config value manually?

**Comment by 132ikl** [comment]
Maybe the default value of `$env.config.use_ansi_coloring` could be set by the envvars? So:

* `$env.config.use_ansi_coloring` is set
  * Use that
* `$env.config.use_ansi_coloring` is unset (not sure how `CLI_COLOR` is supposed to work)
  * If `FORCE_COLOR` is set: `$env.config.use_ansi_coloring` defaults to true
  * Else, if `NO_COLOR` is set: `$env.config.use_ansi_coloring` defaults to false
  * Else, `$env.config.use_ansi_coloring` defaults to auto

**Comment by cptpiepmatz** [comment]
> Maybe the default value of `$env.config.use_ansi_coloring` could be set by the envvars? So:
> 
> * `$env.config.use_ansi_coloring` is set
>   
>   * Use that
> * `$env.config.use_ansi_coloring` is unset (not sure how `CLI_COLOR` is supposed to work)
>   
>   * If `FORCE_COLOR` is set: `$env.config.use_ansi_coloring` defaults to true
>   * Else, if `NO_COLOR` is set: `$env.config.use_ansi_coloring` defaults to false
>   * Else, `$env.config.use_ansi_coloring` defaults to auto

Having the env variables set the config values feels a bit too magic for my taste. Also that would add complexity when adding env variables as you might want to set `NO_COLOR` in the REPL. To make using that feature easiert, we probably need some command the returns the result of the `get` method here.

**Comment by cptpiepmatz** [comment]
> I'm fine with landing this as is but I'm not sure we should force 3rd party authors to understand NO_COLOR, FORCE_COLOR, CLI_COLOR. I could see them respecting $env.config.use_ansi_coloring though. But maybe your code is enough with the auto setting. 🤔

I opened #14683 to solve that.

**Pull request merged**